### PR TITLE
Support for installing CiviCRM-Backdrop via "setup" UI

### PR DESCRIPTION
**Pull request by totten**
Overview
----------

This adds support for installing CiviCRM-Backdrop through the [Setup API](https://docs.civicrm.org/dev/en/latest/framework/setup/new-installer/). Most notable changes:

* Better layout and more useful options on the installation screen
* Extensible installation logic - under the hood, this installer can initialize settings, toggle extensions, etc.
* Reusable installation logic - can run the same installation code for CLI and web UI
* No need to bypass regular module activation or bypass router (no need for `install/index.php` entry-point)

For screenshot and more complete description, see https://github.com/civicrm/civicrm-backdrop/pull/121

### CODE REVIEW

**Comment by colemanw** [comment]
For the 5 people in the world who use Backdrop + Civi, this will be great :+1:

**Comment by eileenmcnaughton** [comment]
@herbdool are you able to review this?

**Comment by totten** [comment]
Why should D7 get all the good patches first? Can't Backdrop get dibs now and then? 😉 

More seriously, this update should also go to D7. I just figured there was more upside starting from BD POV (eg there are some little bits of the existing `install/index.php` that don't work as nicely on BD -- like autodetecting DSN. I'd rather switch over the improved UI than try to fix those.)

**Comment by herbdool** [comment]
I'll test this patch in the next day or so.

**Comment by herbdool** [comment]
So far so good.

I added the two patches to the latest tarball and it installed for me. I had wiped an existing test db but forgot to remove the civicrm.settings.php file which it didn't like. Once I removed that and tried again it ran more or less fine.

I'm also going to try test it on a different host as well since there's a long-standing issue with installing Civi in Lando (at least with Backdrop) which is preventing the clean URLs. Hope to have a cleaner test.

**Comment by herbdool** [comment]
I've got some more feedback. Not sure it's best here or in the other PR on civicrm-backdrop.

tldr; Mostly works except for a handful of odd layout issues.

On a fresh Backdrop site (with a different host) I enable CiviCRM module and see this:

Screenshot from 2020-07-07 11-42-32

Looks mostly fine except for the fact it messed up the module page width. Not sure why Civi is affecting this layout (temporarily since future visits to the page look fine).

After configuring/installing it goes to:

Screenshot from 2020-07-07 09-40-00

It's trying to load http://WEBSITE.org/template.css which doesn't exist at that path.

Then I click "Configure" which brings me to this, which also has the width messed up and the repeated `backdrop_set_message`s are unstyled. And the Civi menu doesn't appear:

Screenshot from 2020-07-07 09-42-01

I don't see console errors but the unstyled messages don't have the proper HTML and classes. They're just being returned as `<li><p class="error">`.

I believe the issue with the width is that the HTML is being inserted outside of the `<div class="l-wrapper-inner">`.

Once I click to the theme menu, the Civi theme settings are already set correctly. I save and then go back to Civi and it looks fine:

Screenshot from 2020-07-07 11-53-18

**Comment by totten** [comment]
@herbdool Thanks for testing! I've pushed a few fixes (for both PRs in `civicrm-core` and `civicrm-backdrop`).

The status-message had a mismatched `</div>`, which I think is what distorted the layout in most of the screenshots. This should be fixed.

Additionally, the status-message was a bit over-zealous - e.g. it was easy to wind up with multiple copies of the status-message in the queue, and some copies of the message may not display to a user until *after* installation. (That leads to confusing messages -- where you've just installed, and it asks you to install again). So I moved the bit which fires the message - instead of firing on every failed call to `civicrm_initialize()`, it fires in `hook_enable` (one-off/immediate message) and `hook_requirements` (a more persistent message - with a red bubble in the navbar).

Screen Shot 2020-07-07 at 7 12 41 PM

On the installer's "finish" page, it was indeed mis-loading `template.css`. I've also pushed up a fix for this - at least, sort of. The CSS file loads. And there's more opportunity to change the appearance of that page... but it's still the same ugly post-install screen that it's always been. So, I don't think it's a regression or needs to be part of this PR... but the post-install screen should be rewritten.

**Comment by herbdool** [comment]
That looks much better!

Screenshot from 2020-07-08 13-27-49

It still shows "Complete setup" as an error in the Civi notifications, but not sure if that's a big deal.

--
I noticed another issue but might not be a regression. I had decided to unselect all Components and it gave back an error "System must have at least one active component", but no way to go back and select a component. If it's not a regression I can add a new issue. I suggest there be some validation on the form to prevent someone from moving forward.

**Comment by totten** [comment]
Oh, that "Complete Setup" message is kind of funny. :) Mitigation: #17776

The validation of "Components" is a pre-existing issue. (If it's easy, I'll post a PR today. Otherwise, yeah, we should do an issue.)

**Comment by totten** [comment]
@herbdool I split off the "Compoments" validation as #17778.

**Comment by seamuslee001** [comment]
I'm going to merge this and the backdrop PR based on @herbdool 's testing
